### PR TITLE
Fix numericality validator to still use value before type cast except Active Record

### LINKED ISSUE #33651: Numericality validation regression in 5.2.1
### Steps to reproduce

https://gist.github.com/zarqman/075179677c21c7aa58d7a7ba3376bac9

### Expected behavior

`validates :quantity, numericality: {greater_than: 0}` should return `/not a number/ when given a value of "invalid".

This is how it worked up through 5.2.0.

### Actual behavior

In 5.2.1, non-numeric user input, such as "invalid" is silently morphed to 0 (numeric zero) and then validated as such, incorrectly returning `/must be greater than 0/` instead.

### System configuration
**Rails version**: 5.2.1

**Ruby version**: 2.5.1

**Mongoid version**: 7.0.1

### Other notes

Because mongoid relies on `before_type_cast` and does not implement `came_from_user?`, the change to eliminate use of `before_type_cast` in Rails 5.2.1 is a breaking change. 

A post-5.2.1 commit (#33603) addresses a different regression from the change in 5.2.1, but does not resolve this one. Additionally, that commit might be incorrect in that it uses `read_attribute(attr_name)` as a fallback, preempting use of `value` which was produced using `read_attribute_for_validation(attr_name)` (which would seem preferable since `raw_value` is being used for validation). It's possible `read_attribute_for_validation` itself needs to fallback to `read_attribute` instead of `send` (whether in ActiveModel or ActiveRecord, I don't know).

It's possible this is a mongoid issue (although since it broke in the Rails 5.2.1 patch-level release, it's arguably still a Rails issue, at least until 6.0).

Also filed Mongoid bug https://jira.mongodb.org/browse/MONGOID-4604

### PR DESCRIPTION
**Pull request by kamipo**
The purpose of fe9547b is to work type casting to value from database.

But that was caused not to use the value before type cast even except
Active Record.

There we never guarantees that the value before type cast was going to
the used in this validation, but we should not change the behavior
unless there is some particular reason.

To restore original behavior, still use the value before type cast if
`came_from_user?` is undefined (i.e. except Active Record).

Fixes #33651.

### CODE REVIEW

File: activemodel/lib/active_model/validations/numericality.rb [hunk 1]
```diff
@@ -21,10 +21,14 @@ def check_validity!
       def validate_each(record, attr_name, value)
         came_from_user = :"#{attr_name}_came_from_user?"
 
-        if record.respond_to?(came_from_user) && record.public_send(came_from_user)
```
Comment (atif-ali) [inline activemodel/lib/active_model/validations/numericality.rb]: asf

**Comment by edzhelyov** [comment]
While this might close #33651 it won't affect the regression that I [have](https://github.com/rails/rails/issues/33651) when using `ActiveAttr`. 

This is basically what I have:
```
class User
  include ActiveAttr::Model

  attribute :account_crn, type: Integer
end

describe User do
  it { should validate_numericality_of(:account_crn) }

  it 'validates account CRN numericality' do
    user.account_crn = 1.1

    user.validate

    user.errors.should be_added :account_crn, :not_an_integer
  end
end
```

The `shoulda-matchers` test is broken and my custom one. I will work out to provide a runnable example tomorrow.

**Comment by cgriego** [comment]
For maximum backwards compatibility, if `#{attr_name}_came_from_user?` isn't defined then ActiveModel should use `#{attr_name}_before_type_cast` like it did before instead of `read_attribute_before_type_cast`.

The `read_attribute_before_type_cast` method is an ActiveRecord interface, but the ActiveModel convention (using `ActiveModel::AttributeMethods`) is that `#{attr_name}_before_type_cast` is the public interface backed by an `attribute_before_type_cast` method. ActiveRecord makes `attribute_before_type_cast` private and exposes its own `read_attribute_before_type_cast` publicly in addition to `#{attr_name}_before_type_cast`.

**Comment by kamipo** [comment]
You are right. I've restored to use `#{attr_name}_before_type_cast` in this method.

**Review by rafaelfranca — APPROVED** [review]

**Pull request merged**